gst/tensor_ros_sink/tensor_ros_sink.c

```diff
@@ -0,0 +1,849 @@
+{
+  GstStructure *structure;
+
+  g_return_val_if_fail (config != NULL, FALSE);
+
+  structure = gst_caps_get_structure (caps, 0);
+
+  if (!gst_structure_has_name (structure, "other/tensor") &&
+      !gst_structure_has_name (structure, "other/tensors")) {
+    err_print ("The input caps, %s, "
+        "do not match the capabilities of the sink pad\n",
+        gst_structure_get_name (structure));
+    return FALSE;
+  }
+
+  gst_tensors_config_from_structure (config, structure);
+
+  return gst_tensors_info_validate (&config->info);
+}
+
+/**
+ * @brief Funtion for new caps.
+ *
+ * GstBaseSink method implementation.
+ */
+static gboolean
+gst_tensor_ros_sink_set_caps (GstBaseSink * sink, GstCaps * caps)
+{
+  GstTensorRosSink *self;
+  GstTensorsConfig in_conf;
+
+  self = GST_TENSOR_ROS_SINK (sink);
+
+  g_mutex_lock (&self->mutex);
+  gst_caps_replace (&self->in_caps, caps);
+  g_mutex_unlock (&self->mutex);
+
+  if (!gst_tensor_ros_sink_read_caps (self->in_caps, &in_conf)) {
+    err_print ("Failed to read the input caps\n");
+    return FALSE;
+  }
+
+  if (!gst_tensors_config_validate (&in_conf)) {
+    err_print ("Failed to validate the input caps\n");
+    return FALSE;
+  }
+
+  self->in_config = in_conf;
+  nns_ros_bridge_set_pub_topic (self->nns_ros_bind_instance, &self->in_config);
+
+  return TRUE;
+}
+
+/**
+ * @brief Funtion to return caps of subclass.
+ *
+ * GstBaseSink method implementation.
+ */
+static GstCaps *
+gst_tensor_ros_sink_get_caps (GstBaseSink * sink, GstCaps * filter)
+{
+  GstTensorRosSink *self;
+  GstCaps *caps;
+
+  self = GST_TENSOR_ROS_SINK (sink);
+
+  g_mutex_lock (&self->mutex);
+  caps = self->in_caps;
+
+  if (caps) {
+    if (filter) {
+      caps = gst_caps_intersect_full (filter, caps, GST_CAPS_INTERSECT_FIRST);
+    } else {
+      gst_caps_ref (caps);
+    }
+  }
+  g_mutex_unlock (&self->mutex);
+
+  return caps;
+}
+
+/**
+ * @brief Handle buffer data.
+ * @return None
+ * @param self pointer to GstTensorRosSink
+ * @param buffer pointer to GstBuffer to be handled
+ */
+static gboolean
+gst_tensor_ros_sink_render_buffer (GstTensorRosSink * self, GstBuffer * inbuf)
+{
+  GstTensorMemory in_tensors[NNS_TENSOR_SIZE_LIMIT];
+  GstMapInfo in_info[NNS_TENSOR_SIZE_LIMIT];
+  GstMemory *in_mem[NNS_TENSOR_SIZE_LIMIT];
+  guint signal_rate;
+  guint num_in_tensors;
+  guint i;
+
+  gboolean notify = FALSE;
+  GstClockTime now = GST_CLOCK_TIME_NONE;
+
+  g_return_val_if_fail (GST_IS_TENSOR_ROS_SINK (self), FALSE);
+
+  num_in_tensors = gst_buffer_n_memory (inbuf);
+
+  for (i = 0; i < num_in_tensors; ++i) {
+    in_mem[i] = gst_buffer_peek_memory (inbuf, i);
+    gst_memory_map (in_mem[i], &in_info[i], GST_MAP_READ);
+
+    in_tensors[i].data = in_info[i].data;
+    in_tensors[i].size = in_info[i].size;
+    in_tensors[i].type = self->in_config.info.info[i].type;
+  }
+
+  g_return_val_if_fail (nns_ros_bridge_publish (self->nns_ros_bind_instance,
```

Comment:
need gst_memory_unmap() when failed.

Reply:
@jaeyun-jung gst_memory_unmap() is added to the case that nns_ros_bridge_publish is failed.